bakery-src/scripts/gdocify_book.py

```diff
@@ -102,26 +102,112 @@ def _convert_cmyk2rgb_embedded_profile(img_filename):
+                                    stdout=asyncio.subprocess.PIPE,
+                                    stderr=asyncio.subprocess.PIPE)
+                                stdout, stderr = await extractembedded.communicate()
+                                # was there an embedded icc profile?
+                                if extractembedded.returncode == 0 and \
+                                        profile.is_file() and \
+                                        profile.stat().st_size > 0:
+                                    cmd = _convert_cmyk2rgb_embedded_profile(
+                                        img_filename)
+                                    print('Convert CMYK (embedded) to '
+                                          'RGB: {}'.format(img_filename))
+                                else:
+                                    cmd = _convert_cmyk2rgb_no_profile(
+                                        img_filename)
+                                    print('Convert CMYK (no profile) to '
+                                          'RGB: {}'.format(img_filename))
+                                if profile.is_file():
+                                    profile.unlink()  # delete file
+                        else:
+                            cmd = _universal_convert_rgb_command(img_filename)
+                            print('Warning: Convert exceptional color '
+                                  'space {} to RGB: {}'.format(colorspace, img_filename))
+                        # convert command itself
+                        fconvert = await asyncio.create_subprocess_shell(
+                            cmd, stdout=asyncio.subprocess.PIPE,
+                            stderr=asyncio.subprocess.PIPE)
+                        stdout, stderr = await fconvert.communicate()
+                        if fconvert.returncode != 0:
+                            raise Exception('Error converting file {}'.format(img_filename) +
+                                            ' to RGB color space: {}'.format(stderr))
+            except UnidentifiedImageError:  # pragma: no cover
+                # do nothing if we cannot open the image
+                print('Warning: Could not parse JPEG image with PIL: ' + str(img_filename))
+    else:
+        raise Exception('Error: Resource file not existing: ' + str(img_filename))  # pragma: no cover
+
+
+class AsyncJobQueue:
+    def __init__(self, worker_count, qsize=None):
+        self.worker_count = worker_count
+        self.queue = (asyncio.Queue(qsize) if qsize is not None
+                      else asyncio.Queue())
+        self.workers = []
+
+    async def __aenter__(self):
+        async def worker(queue):
+            while True:
+                try:
+                    job = await queue.get()
+                    await job
+                    queue.task_done()
+                except Exception as e:  # pragma: no cover
```

Comment:
Of the set of changes this was the only area that kinda bothered me. The fact that we bail on the first error instead of processing them all and reporting those that failed (like using `run_until_complete()`. I tried investigating solutions but there wasn't anything I could find that I liked better other than this [solution](https://www.pythonfixing.com/2022/03/fixed-python-asyncio-queuejoin-finishes.html).

Reply:
Yes, I agree. Normally I would have two-way communication and have each task respond, via a second queue, with an object that describes any errors that occurred. That would allow more control over the way errors are handled.

In this case, however, the script bailed out on the first error before these changes were made. I wanted to keep error handling similar to how it was before.